src/psyclone/psyir/backend/fortran.py

```diff
@@ -405,6 +338,72 @@ def __init__(self, skip_nodes=False, indent_string="  ",
                                             indent_string,
                                             initial_indent_depth,
                                             check_global_constraints)
+        # Reverse the Fparser2Reader maps that are used to convert from
+        # Fortran operator names to PSyIR operator names.
+        self._operator_2_str = {}
+        self._reverse_map(self._operator_2_str,
+                          Fparser2Reader.unary_operators)
+        self._reverse_map(self._operator_2_str,
+                          Fparser2Reader.binary_operators)
+        self._reverse_map(self._operator_2_str,
+                          Fparser2Reader.nary_operators)
+
+    @staticmethod
+    def _reverse_map(reverse_dict, op_map):
+        '''
+        Reverses the supplied fortran2psyir mapping to make a psyir2fortran
+        mapping. Any key that does already exist in `reverse_dict`
+        is not overwritten, only new keys are added.
+
+        :param reverse_dict: the dictionary to which the new mapping of \
+            operator to string is added.
+        :type reverse_dict: dict from \
+            :py:class:`psyclone.psyir.nodes.BinaryOperation`, \
+            :py:class:`psyclone.psyir.nodes.NaryOperation` or \
+            :py:class:`psyclone.psyir.nodes.UnaryOperation` to str
+
+        :param op_map: mapping from string representation of operator to \
+                       enumerated type.
+        :type op_map: :py:class:`collections.OrderedDict`
+
+        '''
+        for operator in op_map:
+            mapping_key = op_map[operator]
+            mapping_value = operator
+            # Only choose the first mapping value when there is more
+            # than one.
+            if mapping_key not in reverse_dict:
+                reverse_dict[mapping_key] = mapping_value.upper()
+
+    def is_intrinsic(self, operator):
```

Comment:
IMHO the original design is very bad:
1) It pollutes the namespace by adding some random global functions (what happens if another module defines a function with the same name? Might then the result depend on the order in which imports happen???). That's why I moved it into the class (originally as static function).
2) But even then, the bad design is that the instance of the Fortran writer depends on global variables from a different module. So you can create an instance, use the function, and it works. But them 'something' modifies a totally unrelated variable in a different module, and suddenly this class could break. Seriously ... what is it with all this global stuff everywhere?  :)   IMHO this function should copy the values when the instance is created, and then use its own copy, to be truly self contained. But since this was not the focus of this PR, and I did not want to interfere with existing code, I did not fix these design problems.
3) The SymPy writer does it properly - creating the data in the constructor, and then be self-contained. But that of course requires `self`. And I think it's better to be consistent with the various function, so I added the `self` in the Fortran base class, (and added a directive to ignore the pylint warning).

I am happy to also fix this  - I am just trying to be good and keep PRs small 😊

Reply:
Those sound like good points. Perhaps we should have a separate issue to remove the global-variable dependence from FortranWriter.